src/canisters/frontend/ic-certified-assets/src/state_machine.rs

```diff
@@ -830,117 +999,158 @@ impl From<StableState> for State {
-            return;
-        }
-    }
-
-    // No known encodings found. Just pick the first one. The exact
-    // order is hard to predict because we use a hash map. Should
-    // almost never happen anyway.
-    if let Some(enc) = asset.encodings.values_mut().next() {
-        asset_hashes.insert(key.to_string(), enc.sha256);
-        for dependent in dependent_keys {
-            asset_hashes.insert(dependent, enc.sha256);
-        }
-        enc.certified = true;
-    }
-}
-
-fn witness_to_header(witness: HashTree, certificate: &[u8]) -> HeaderField {
-    use ic_certified_map::labeled;
-
-    let hash_tree = labeled(b"http_assets", witness);
-    let mut serializer = serde_cbor::ser::Serializer::new(vec![]);
-    serializer.self_describe().unwrap();
-    hash_tree.serialize(&mut serializer).unwrap();
-
-    (
-        "IC-Certificate".to_string(),
-        String::from("certificate=:")
-            + &base64::encode(certificate)
-            + ":, tree=:"
-            + &base64::encode(&serializer.into_inner())
-            + ":",
-    )
-}
-
-fn merge_hash_trees<'a>(lhs: HashTree<'a>, rhs: HashTree<'a>) -> HashTree<'a> {
-    use HashTree::{Empty, Fork, Labeled, Leaf, Pruned};
+    asset.update_ic_certificate_expressions();
+
+    // Order of encodings is relevant for v1. Follow ENCODING_CERTIFICATION_ORDER,
+    // then follow the order of existing encodings.
+    // For v2, it is important to certify all encodings, therefore all encodings are added to the list.
+    let mut encoding_order: Vec<String> = ENCODING_CERTIFICATION_ORDER
+        .iter()
+        .map(|enc| enc.to_string())
+        .collect();
+    encoding_order.append(&mut asset.encodings.keys().map(|s| s.into()).collect());
+
+    let mut keys_to_insert_hash_for = dependent_keys;
+    keys_to_insert_hash_for.push(key.into());
+    let keys_to_insert_hash_for: Vec<_> = keys_to_insert_hash_for
+        .into_iter()
+        .map(|key| {
+            let v1_hash_path = AssetPath::from(&key).asset_hash_path_v1();
+            (key, v1_hash_path)
+        })
+        .collect();
+
+    // Insert certified response values into hash_tree
+    // Once certification v1 support is removed, encoding_order.iter() can be replaced with asset.encodings.keys()
+    for enc_name in encoding_order.iter() {
+        let Asset {
+            content_type,
+            encodings,
+            max_age,
+            headers,
+            ..
+        } = asset;
+        if let Some(enc) = encodings.get_mut(enc_name) {
+            let mut encoding_headers: Vec<(String, Value)> = build_headers(
+                headers.as_ref().map(|h| h.iter()),
+                max_age,
+                &*content_type,
+                enc_name,
+                enc.ic_ce.as_ref().map(|ce| &ce.ic_certificate_expression),
+            )
+            .into_iter()
+            .map(|(k, v)| (k, Value::String(v)))
+            .collect();
+
+            let new_response_hashes = {
```

Comment:
Can we decompose this method into some smaller ones?  This block might be a starting point.

Reply:
I was able to extract a lot. It is much easier to read now